Revamped iomgr's callback mechanism

dgquintas opened the pull request:
This is a tentative (see discussion below) pull request for the modification of the way the iomgr deals with callbacks.

In particular, the allocation of the callback (now exposed and called grpc_iomgr_cb) is the responsibility of the iomgr user. In some cases, this allows the user to reduce the number of mallocs and/or group them into single chunk.

**Discussion**: In instances where it isn't possible to anticipate the number of necesary callback allocations,  the callback needs to be allocated "in situ" right before passing it over to the iomgr. The release of these allocations is currently performed conditionally inside iomgr (depending on whether the callback has been marked as "externally managed"). 
The alternative is to make the deallocation part of the callback itself. However, this introduces a non trivial amount of complexity due to not always having control over the callback function. For instance, fd_posix.c:notify_on creates a callback whose function comes from the outside (closure->cb). In other cases, even when we have access to the callback function, such as credentials.c:on_simulated_token_fetch_done, they'd have to be modified to receive a pointer to its enclosing grpc_iomgr_cb. 
Both these scenarios can be addressed by creating wrapper around callbacks that'd run whichever callback function and then free itself. However, a functionally equivalent approach is have the iomgr conditionally free(), as described.

ctiller commented on src/core/iomgr/iomgr.h:
```diff
@@ -37,11 +37,22 @@
+  grpc_iomgr_cb_func cb;
+  void *cb_arg;
+  int success;
+  int is_ext_managed;  /** is memory being managed externally? */
```
So I'd like to make this field disappear before we submit this.

ctiller replied:
(I'll help though)

dgquintas replied:
Ack, I figured. The reason for its existence is in the pull req. description. Let's chat in person.

ctiller commented on src/core/iomgr/iomgr.h:
```diff
@@ -37,11 +37,22 @@
 /* gRPC Callback definition */
 typedef void (*grpc_iomgr_cb_func)(void *arg, int success);
 
+typedef struct grpc_iomgr_cb {
```
Let's rename this grpc_iomgr_closure, and remove that type from fd_posix.h.

They both represent a callback paired with a parameter.
This makes the notify_on case in fd_posix.c actually trivial, and removes a use of is_ext_managed.

ctiller commented on src/core/iomgr/fd_posix.c:
```diff
@@ -115,7 +121,10 @@ static void unref_by(grpc_fd *fd, int n) {
   if (old == n) {
     close(fd->fd);
-    grpc_iomgr_add_callback(fd->on_done, fd->on_done_user_data);
+    fd->on_done_iocb->cb = fd->on_done;
```
Mutate the interface to take a grpc_iomgr_closure and push the storage problem up one layer?

dgquintas replied:
Remains to be done. I'll take care of it tomorrow.

dgquintas replied:
Done (albeit differently, as discussed offline)

ctiller commented on src/core/iomgr/fd_posix.c:
```diff
@@ -172,20 +181,25 @@ void grpc_fd_ref(grpc_fd *fd) { ref_by(fd, 2); }
 void grpc_fd_unref(grpc_fd *fd) { unref_by(fd, 2); }
 
 static void make_callback(grpc_iomgr_cb_func cb, void *arg, int success,
```
change signature to take a grpc_iomgr_closure* instead of cb, arg, and then call it directly if allow_synchronous_callback?

ctiller commented on src/core/iomgr/iomgr.h:
```diff
@@ -37,11 +37,22 @@
+  struct grpc_iomgr_cb *next;  /** Do not touch */
+} grpc_iomgr_cb;
+
+grpc_iomgr_cb *grpc_iomgr_cb_create(grpc_iomgr_cb_func cb, void *cb_arg,
```
We ultimately won't need this routine.

However, consider providing a grpc_iomgr_closure_init(grpc_iomgr_closure *, ...) to save line count across the project.

dgquintas replied:
Agreed. This was actually on my TODO but forgot to make the change.

ctiller commented:
So I think I dropped a suggestion or two that lets us remove the is_ext_managed thing completely - take a look and let me know what you think. If there's a case that's troubling and I didn't comment on it, then probably I missed it this morning and flag it to me.

dgquintas commented:
PTAL. Addressed all comments except for one that fell through the cracks (to be had for breakfast tomorrow).

All tests pass, incluiding "asan" (inasmuch it fails no more than HEAD :))

(also, I'm not sure why github talks about "outdated diffs". I haven't rebased, hmm).

ctiller commented on src/core/iomgr/fd_posix.c:
```diff
@@ -212,8 +214,8 @@ static void notify_on(grpc_fd *fd, gpr_atm *st, grpc_iomgr_closure *closure,
       gpr_atm_rel_store(st, NOT_READY);
-      make_callback(closure->cb, closure->cb_arg,
-                    !gpr_atm_acq_load(&fd->shutdown),
+      closure->success = -1;
```
What does -1 mean here?

ctiller replied:
(we usually use 0=false, 1=true)

dgquintas replied:
Removed. Leftover.

ctiller commented on src/core/iomgr/fd_posix.c:
```diff
@@ -258,25 +260,31 @@ static void set_ready(grpc_fd *fd, gpr_atm *st,
   gpr_mu_unlock(&fd->set_state_mu);
   success = !gpr_atm_acq_load(&fd->shutdown);
-  make_callbacks(&cb, ncb, success, allow_synchronous_callback);
+  assert(ncb <= 1);
```
GPR_ASSERT()

dgquintas replied:
Done. Changed a couple other assert()s in the file while at it.

ctiller commented on src/core/iomgr/fd_posix.c:
```diff
@@ -258,25 +260,31 @@ static void set_ready(grpc_fd *fd, gpr_atm *st,
      notify_on) */
   int success;
-  grpc_iomgr_closure cb;
+  grpc_iomgr_closure closure;
```
What if this was:
```
grpc_iomgr_closure *closure;
```

and set_ready_locked took a `grpc_iomgr_closure**` - so it could simply set the pointer that we already have, and we'd not need to do an allocation here either?

dgquintas replied:
Done.

ctiller commented on src/core/surface/server.c:
```diff
@@ -415,20 +419,29 @@ static void server_on_recv(void *ptr, int success) {
     case GRPC_STREAM_CLOSED:
       gpr_mu_lock(&chand->server->mu);
       if (calld->state == NOT_STARTED) {
+        grpc_iomgr_closure *kill_zombie_closure =
```
We can guarantee that there's at most one kill_zombie outstanding at once - meaning we can probably squeeze this onto call_data.

It's probably going to be bad for memory usage, as this mechanism is used rarely. 
However: it lets us get rid of managed closures entirely, and when we get to memory usage optimization, maybe something presents itself as the obvious thing to do.

dgquintas replied:
Done.

ctiller commented on src/core/security/credentials.c:
```diff
@@ -871,9 +871,12 @@ static void fake_oauth2_get_request_metadata(grpc_credentials *creds,
-        on_simulated_token_fetch_done,
+    grpc_iomgr_closure *on_simulated_token_fetch_done_closure =
+        gpr_malloc(sizeof(grpc_iomgr_closure));
+    grpc_iomgr_managed_closure_init(
```
This might take an iteration or two with @jboeuf - but I think the changes we were talking about yesterday evening might help here.

I'm worried that we're going to spam auth servers if we are under load with the current architecture. Imagine running 10kqps and having an auth server for token fetches with 10ms latency (which is not at all unbelievable) - now we're doing 100 identical requests against the auth server, and not making that latency any better.

I'd prefer to see something in either auth.c or credentials.c that guarantees that at most one request for new credential tokens is made at any given time. If we can get that, then this path gets a little cleaner in that we can preallocate the closure object as we do in the rest of the codebase.

ctiller replied:
We should chat about this one.

dgquintas replied:
This remains the single usage of the managed_closure concept. Course of action TBD, added TODO.

dgquintas commented:
Comments addressed. Thanks! PTAL.

Two things remain: 
- The single use of managed_closure in security/credentials.c 
- Addition of comments if the newly introduced code looks good.

dgquintas commented:
@nicolasnoble Agreed. Will do once the code changes behind this pull req. crystallize.

dgquintas commented:
Ok, this pull req. should be all ready now. I've included some comments in iomgr.h. @ctiller could you pleasemake sure they are accurate?

dgquintas commented:
Green! (modulo stuck coveralls)

ctiller commented:
Looks good. Why don't we kill the to-do in credentials while we are here (should only be a small refactor).

dgquintas commented:
Done.

dgquintas commented:
*Actually* done this time :)

dgquintas commented:
Green!
